src/util/amount.rs

```diff
@@ -995,6 +995,184 @@ pub mod serde {
     }
 }
 
+/// Type for dealing with unknown amounts
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[derive(Clone, Copy, PartialEq, PartialOrd, Debug)]
+pub enum CoinAmount {
```

Comment:
Hmm, is roundtripping for this important? There is already a serde-related module in `util::amount` that does flexible serialization with sats or btc. It's possible to add a deserializer that can have both, but it's harder to then roundtrip that, so you'd have to chose one to serialize with.

How about something like this: https://github.com/rust-bitcoin/rust-bitcoin/pull/509

Reply:
Idk what roundtripping means in this cirumstance. Do you mean the From impls?

The issue with the flexible serialization is that it isn't actually flexible and requires you to annotate with a serializer anywhere you use that type, and then the selection of schemars schema is also not type safe.

A type like CoinAmount can be used in a publicly facing API that can accept either a Bitcoin or a Satoshis amount.

Reply:
Hmm, yeah I don't know how the different serde deserializers interact with the schema mechanics.

I'm personally not the biggest fan to catering our internal types to the likes of very specific serialization methods, but always building the serialization support around the existing types. That's what has been done with the `Amount` type. It makes sense internally and there has been provided some infrastructure to use it with serde. Ideally if there are difficulties with schemas, they'd be solved by providing functionality around the `Amount` and `SignedAmount` types instead of given a new type.

There's various reasons for this, one is that we want to have a single type per concept, that all serializations can use, not a type per serialization. I can imagine that the `CoinAmount` doesn't work with other serialization methods like protobuf or so. Also, we shouldn't have types reflect serialization-specific information. Meaning that for the sake of the type itself, it's not useful to keep track on whether the value was derived from serializing `0.01 BTC` or `1000000 sat`, which the enum seems to do.

What I meant with roundtripping and https://github.com/rust-bitcoin/rust-bitcoin/pull/509 is that using the flexible deserializer in https://github.com/rust-bitcoin/rust-bitcoin/pull/509, when re-serializing, the original serialization will be lost.

Ultimately, I think if a specific implementation needs a specific serialization of amounts, it's quite trivial to quickly (1) wrote a serde serialization module around `Amount` or (2) write a newtype for it and use that for the API of that implementation.

Reply:
It just seems remarkably easy to, for a programmer, mess up an incoming type as sats or btc. I think an explicitly tagged enum is a much better API for all users as it is completely self-documenting and unambiguous. This isn't just an issue generating the schemas, it's just odd to me that we don't have a type-level way of disambiguatiing the amounts.

Another alternative would be to make Amounts split into AmountsSats and AmountsBtc and then you can use Either<AmountsSats, AmountsBtc> (or result...) to express an api taking either.

Reply:
But, there is only 1 amount "concept", right? "an amount", aka `CAmount` in Core.. BTC is just a denomination. And denominations only matter when you're _representing_ the amount. And representation is an application-specific thing, so I think providing flexible tools to help _represent_ our amount type easily in various ways is the way to go instead of creating different types for different presentations.

Reply:
I think I've figured out the right API here -- what we probably want is to have the type be Amount and use try_from for Amount and serialize_with that defaults to CoinAmount::Sats wrapper. That way the type in the structs is an amount, but we can handle inbound of either.

Then if you need the *specific encoding* of Amount as a plain u64 or float (without tag), you need to ask for it with an override to the default behavior.

(I also think we should unify Signed and Unsigned amount as CAmount is signed, but that's another question).

In short: at the type level, have a single simple newtype. At the serialization level, have explicit tagged representations with a tagged representation and a fallback mechanism. I'll need to think through the deduction rules and type them up

Reply:
Hmm yeah I'm not a fan of that approach, but if others disagree, that's fine.

As for Amount and SignedAmount, I think it's useful to have amounts of which you know they cannot be negative.